spec/unit/lcm/lcm2_spec.rb

```diff
@@ -45,10 +37,111 @@
+          GDC_GD_CLIENT: client,
+          GDC_LOGGER: logger,
+          domain: domain
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'skips actions in skip_actions' do
+        expect(GoodData::LCM2::CollectSegments).not_to receive(:call)
+        expect(GoodData::LCM2::SynchronizeUsers).not_to receive(:call)
+        GoodData::LCM2.perform('users', params)
+      end
+    end
+
+    context 'when mandatory params are given and hello action is performed' do
+      let(:params) do
+        params = {
+          GDC_LOGGER: logger,
+          message: 'Ahoj'
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'should finish successfully' do
+        result = GoodData::LCM2.perform('hello', params)
+        pp result
+        expect(result[:actions]).to eq(['HelloWorld'])
+        expect(result[:results]["HelloWorld"][0][:message]).to eq('Ahoj')
+        expect(result[:params][:message]).to eq('Ahoj')
+        expect(result[:params][:gdc_logger]).to eq(logger)
+        expect(result[:params][:iterations]).to eq(1)
+      end
+    end
+
+    context 'when invalid mode is specified' do
+      let(:params) do
+        params = {
+          GDC_LOGGER: logger
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'should throw exception' do
+        expect do
+          GoodData::LCM2.perform('invalid', params)
+        end.to raise_error { |e| expect(e.message).to include('Invalid mode specified \'invalid\'') }
+      end
+    end
+
+    [true, false].each do |fail_early|
+      [true, false].each do |strict_mode|
+        context "when fail_early is #{fail_early}, strict_mode is #{strict_mode} and error occurs" do
+          let(:params) do
+            params = {
+              GDC_LOGGER: logger,
+              fail_early: fail_early,
+              strict: strict_mode
+            }
+            GoodData::LCM2.convert_to_smart_hash(params)
+          end
+
+          # Test class 1
+          class Action1 < GoodData::LCM2::BaseAction
+            PARAMS = {}
+            DESCRIPTION = "Test action"
+          end
+
+          # Test class 2
+          class Action2 < Action1
+          end
+
+          # Test class 3
+          class Action3 < Action1
+          end
+
+          before do
+            allow(GoodData::LCM2).to receive(:get_mode_actions) { [Action1, Action2, Action3] }
+            expect(Action1).to receive(:call).and_return([{ :a1 => 'a1' }])
+            expect(Action2).to receive(:call).and_raise("boom")
+            if fail_early
+              expect(Action3).not_to receive(:call)
+            else
+              expect(Action3).to receive(:call).and_return([{ :a3 => 'a3' }])
+            end
+          end
+
+          it "should fail and #{fail_early ? 'stop' : 'continue'} performing next actions" do
+            if strict_mode
+              expect do
```

Comment:
warning: Similar code found in lcm2_spec.rb:87, lcm2_spec.rb:132 (mass = 34) (flay)